File: week2/happy-number.java
class Solution {
    public boolean isHappy(int n) {
        HashSet<Integer> set = new HashSet<>();
        int sum = n;
        while (sum != 1) {
            set.add(sum);
            sum = sumOfSquares(sum);
            if (set.contains(sum))
                return false;
        }
        return true;
    }
    
    public int sumOfSquares(int n) {
        int sum = 0;
        while (n > 0) {
            sum += (int)(n % 10) * (int)(n % 10);
            n /= 10;
        }
        return sum;
    }
}

File: 2026/week2/happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        s = set()
        current_sum = n
        while current_sum != 1:
            if current_sum in s:
                return False
            s.add(current_sum)
            new_sum = 0
            while current_sum > 0:
                right = current_sum % 10
                new_sum += right * right
                current_sum //= 10
            current_sum = new_sum
        return True
